### ISSUE TITLE
Number field #max values allow settings impossible to save in the database, causing exceptions

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4751]
**Description of the bug**

Number field setting validation is incomplete. First of all, there's currently a hardcoded max value if nothing is set by admins (1000000, core/includes/form.inc). This has been addressed in a forum post, because the help text for the field setting is misleading. But that's not what I'm trying to address here.

The max setting can get overridden via UI. And with an overridden value it's very easy to produce PDOExceptions.

**Steps To Reproduce**

1. Add an integer field to a content type
2. Set Maximum to something high (_higher_ than 2147483648, explanation follows)
3. Turn on display of all errors
4. Add a node of that type and input 2147483649
5. Try to save the node

**Actual behavior**

![sql-exception-range](https://user-images.githubusercontent.com/842066/98817564-8f84bf00-242a-11eb-822b-4da1c5a3f914.png)

EntityStorageException: SQLSTATE[22003]: Numeric value out of range: 1264 Out of range value for column 'field_integer_value' at row 1 in NodeStorageController->save() (line 510 of ...core/modules/node/node.entity.inc).
SQLSTATE[22003]: Numeric value out of range: 1264 Out of range value for column 'field_integer_value' at row 1

**Expected behavior**

Guess what :wink: 

**Additional information**

What's that "magic" number 2147483648?

It's the max value MariaDB can store in signed integer columns. https://mariadb.com/kb/en/int/ It's almost identical in [MySQL](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html) and also PostgreSQL.
For integer fields, the Number module creates database table columns of type `int(11)`

The situation for decimal fields is pretty alike. It's possible to set max values that can't get saved (exceed the allowed precision), we should take database limits into account. For floats it's less dramatic, as MariaDB stores huge ones without any problems.

The exception is only thrown when [strict mode](https://mariadb.com/kb/en/sql-mode/#strict-mode) is set - default from MariaDB 10.2.4.

I'm not sure yet what's the best approach. Very likely it makes sense to prevent admins from setting such an impossible max value (field UI, `number_field_instance_settings_form_validate()`). Would it also make sense to prevent values like that in `form_validate_number()`?

BTW: the error is reproducible on D8.

**Comment by indigoxela** [comment]
I decided to postpone the field widget value validation. It would be blocked by that other decimal field bug and needs some feedback/decision anyway.

Done: prevent admins from setting impossible values for max/min of integer and decimal fields. And I also added some tests for it (also for the min may not be greater than max, which was still pending).

**How to test**

Log into the sandbox and add an integer / decimal field to a content type. Try to set impossible values (see issue description for examples). You shouldn't be able to.

**Comment by klonos** [comment]
Thanks @indigoxela 🙏  ...I've tested with the STR you've provided, side by side with a demo sandbox (currently running 1.17.2) and the PR sandbox, and things work as expected now 👍 

Can we also do the following as part of this ticket:

- change the min/max fields in the field configuration form to be `'type' => 'number'`, with proper `'#min'` and `'#max'` values, instead of using text fields.
- make the default value pick its `'#min'` and `'#max'` values from whatever is set in the "Minimum" and "Maximum" fields respectively.
- either change that hard-coded default of `1000000`, or make sure that we have help text explaining that that's what the default is (it doesn't make sense to leave), or perhaps placeholder text in the field itself?

**Comment by indigoxela** [comment]
@klonos many thanks for testing. :bowing_woman: 

> Can we also do the following as part of this ticket

I'd actually prefer to do it in a follow-up. Let me explain:

> change the "Minimum" and "Maximum" fields in the field configuration form to be 'type' => 'number'

I need to do some more recherche and testing re that. Currently the provided values get saved as string. This has the advantage that we do not have to consider faulty floating point precision. It might be possible, but while comparing to D8, which does have number input items, I discovered a _nasty_ side effect: a provided value of 99999999999999,99 turns into 1.0E+14 - large floats get rounded.

> either change that hard-coded default of 1000000, or make sure that we have help text explaining

That's also something that needs thorough consideration. The core form item probably has that value set because it's mostly used for weight fields (integers). The Number field items have different types (int|float|decimal) which all have different requirements re min/max values.

But most of all: it will be necessary to get the fix from #3792 merged so that people can properly test decimals/floats. The step validation is totally broken ATM.

**Comment by herbdool** [comment]
@indigoxela you mentioned https://github.com/backdrop/backdrop-issues/issues/3792 here. Does this mean this PR needs to be updated or still ready for review?

**Comment by klonos** [comment]
The PR needs a rebase anyway - been a year+ now.

**Comment by indigoxela** [comment]
It's been a while here... :grin:

Rebased, so it's ready for another round of testing/review.

**Comment by indigoxela** [comment]
Another rebase, just in case someone's interested in reviewing.

**Comment by herbdool** [comment]
The code looks to me. Test coverage looks good too. Thanks @indigoxela

**Comment by quicksketch** [comment]
Thanks @indigoxela, @herbdool, and @klonos! Did a final review and looks good to me. I merged https://github.com/backdrop/backdrop/pull/3401 into 1.x and 1.21.x.
